api/v1/clusterwidenetworkpolicy_types.go

```diff
@@ -109,8 +116,86 @@ type EgressRule struct {
 	// empty or missing, this rule matches all destinations (traffic not restricted by
 	// destination). If this field is present and contains at least one item, this rule
 	// allows traffic only if the traffic matches at least one item in the to list.
+	// To rules can't contain ToFQDNs rules.
 	// +optional
 	To []networking.IPBlock `json:"to,omitempty"`
+
+	// List of FQDNs(fully qualified domain name) for outgoing traffic of a cluster for this rule.
+	// Items in this list are combined using a logical OR operation. This field is used as
+	// whitelist of DNS names. If none specified, rule will not be applied.
+	// ToFQDNs rules can't contain To rules.
+	// +optional
+	ToFQDNs []FQDNSelector `json:"toFQDNs,omitempty"`
+}
+
+// FQDNSelector describes rules for matching DNS names.
+type FQDNSelector struct {
+	// MatchName matches FQDN.
+	// +kubebuilder:validation:Pattern=`^([-a-zA-Z0-9_]+[.]?)+$`
+	MatchName string `json:"matchName,omitempty"`
+
+	// MatchPattern allows using "*" to match DNS names.
+	// "*" matches 0 or more valid characters.
+	// +kubebuilder:validation:Pattern=`^([-a-zA-Z0-9_*]+[.]?)+$`
+	MatchPattern string `json:"matchPattern,omitempty"`
+
+	// Sets stores nftables sets used for rule.
```

Comment:
So, does the controller use this field internally only? Or is this managed by the user?

Reply:
It's used only internally. To restore nftables rules in their were reset.

Reply:
Mh... I am not sure if a controller should store state inside the same resource that it is reconciling -- also from the user-facing API it feels a bit unusual. Can't this grow pretty large? Is it so costly to implement this stateless? If not, could placing a JSON database file or something onto to machine be viable solution, too?

Reply:
I guess if someone sets the rule like `*` it can grow pretty large, yes)

I assume that we can use local DB to store this info. Although, there's some benefits to storing it in a resource status -- as user can actually see to what IPs the rule is resolved. Also, monitoring functionality could be grown other time(storing number of requests?). While local DB would be very inconvenient for monitoring purposes.

Reply:
Regarding stateless implementation. I honestly don't remember and can't find why we went with "states caching" approach. But in case of pattern rules, when no specific address is provided, some dropped requests are possible before DNS proxy updates the nftables rules.

Reply:
UPDATE: it looks like it's not used for cleaning after CWNP deleted. So it's purely for restoration.

Reply:
Config map per CWNP resource? One of conveniences of current implementation is if CWNP deleted, all related info deleted automatically and we don't have to do any additional steps. Of course, it's convenience is purely for developers.